[Zheng Yexun, Bryan] iP

Ashuh opened the pull request:
### Duke

> “Life is too complicated not to be orderly.” – Martha Stewart [(source)](https://dansilvestre.com/productivity-quotes/#10_Life_is_too_complicated_not_to_be_orderly_-_Martha_Stewart)

Duke helps you to remember the things you need to do. It is

- Text based
- Easy to learn
- ~~FAST~~ *SUPER* FAST to use
- **FREE**!

## Getting Started

1. Download the latest release from [here](https://github.com/Ashuh/ip/releases/tag/A-Jar).
2. Launch using the following command: `java -jar Duke.jar`.
3. Let Duke manage your tasks for you! :smirk:

## Command Format

- Add a ToDo: `todo /desc return book`
- Add a Deadline: `deadline /desc return book /by 2022-1-29 1230`
- Add an Event: `event /desc return book /at 2022-1-29 1230 /dur 1h15m`
- Delete a task: `delete /id 1`
- Mark a task: `mark /id 1`
- Unmark a task: `unmark /id 1`
- Exit Duke: `bye`
- List tasks: `list`
- Search for a task: `find /key book`

The parameter names for each command are defined in `CommandType.java`. You can change them by simply modifying the following lines of code:

```java
public enum CommandType {
    ADD_TODO("todo", "desc"),
    ADD_DEADLINE("deadline", "desc", "by"),
    ADD_EVENT("event", "desc", "at", "dur"),
    DELETE_TASK("delete", "id"),
    MARK_TASK("mark", "id"),
    UNMARK_TASK("unmark", "id"),
    EXIT("bye"),
    LIST("list"),
    FIND("find", "key");
}
```

## Features

- [x] Managing to do
- [x] Managing deadline
- [x] Managing event
- [x] Data storage
- [x] Searching by description
- [ ] Searching by time (Coming soon)

Commit:
Add Gradle support

Ashuh committed:
Implement greet, echo, exit

Ashuh committed:
Implement add, list

Ashuh committed:
Refactor output formatting

Ashuh committed:
Add ability to mark tasks as done

Ashuh committed:
Add support for tracking todos, deadlines, events

Ashuh committed:
Remove trailing whitespace from logo

Ashuh committed:
Add automated text UI test

Ashuh committed:
Encapsulate user input processing logic in processInput()

Ashuh committed:
Handle invalid commands and missing Task information

Ashuh committed:
Add support for deleting tasks

Ashuh committed:
Use enum to represent types of commands

Ashuh committed:
Implement task list saving and loading

Ashuh committed:
Represent dates and times using LocalDateTime and modify command format

Ashuh committed:
Merge branch 'branch-Level-7'

Ashuh committed:
Merge branch 'branch-Level-8'

Ashuh committed:
Update Ui tests

Ashuh committed:
Extract user interaction logic into Ui class

Ashuh committed:
Use TaskList class to store tasks

Ashuh committed:
Extract task loading and saving logic into Storage class

Ashuh committed:
Extract user input parsing logic into Parser class

Ashuh committed:
Organize classes into packages

Ashuh committed:
Add JUnit tests

Ashuh committed:
Add JavaDoc comments

Ashuh committed:
Rename variables to comply with coding standard

Ashuh committed:
Add ability to search for Tasks by keyword

Ashuh committed:
Merge branch 'branch-A-JavaDoc'

Ashuh committed:
Merge branch 'branch-A-CodingStandard'

Ashuh committed:
Merge branch 'branch-Level-9'

Ashuh committed:
Merge branch 'master' into add-gradle-support

Ashuh committed:
Change mainClassName in build.gradle

Ashuh committed:
Disable checkstyle

Ashuh committed:
Merge branch 'add-gradle-support'

Ashuh committed:
Add checkstyle support

Ashuh committed:
Fix checkstyle errors

Ashuh committed:
Merge branch 'add-checkstyle'

Ashuh committed:
Implement GUI and remove old text-based UI

Ashuh committed:
Add javadoc comments

Ashuh committed:
Display DukeException messages using Alert dialog and preserve description input if an exception is thrown

Ashuh committed:
Merge branch 'branch-Level-10'

MontyPython28 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,121 @@
+import duke.task.TaskList;
+import duke.task.ToDo;
+
+/**
+ * Encapsulates the main high-level logic of the bot.
+ */
```
Maybe you can use the` /** header_comment_here*/` structure for the one-line header comments?

MontyPython28 commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Represents the possible Command types.
+ */
+public enum CommandType {
```
I like how you used enum here!

MontyPython28 commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,57 @@
+    private final String command;
+    private final String[] params;
+
+    CommandType(String command, String... params) {
```
I like how you used varargs here!

MontyPython28 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,54 @@
+        if (id >= size()) {
+            throw new DukeException("Invalid id");
+        }
+
```
Maybe remove this empty line

MontyPython28 commented on src/test/java/duke/command/CommandTypeTest.java:
```diff
@@ -0,0 +1,27 @@
+
+import duke.DukeException;
+
+public class CommandTypeTest {
```
Maybe put header comments for your test methods?

MontyPython28 reviewed (commented):
Looking really good! I could only find really minor issues. But overall, it looks very well done.

Ashuh committed:
Fix Tasks not being updated properly when search function is used

Each TaskCard has an index which is used to access the Task it represents in the TaskList. The index assigned is the position of the TaskCard in the VBox.

When the search function is used, the position of the TaskCard does not match the index of the Task in the TaskList.

Let's perform filtering of Tasks in Window instead of TaskList.

This allows the original index of the filtered Tasks to be tracked and passed to each TaskCard.

Ashuh committed:
Add assertions

Ashuh committed:
Delete old text-based UI code

Ashuh committed:
Merge pull request #1 from Ashuh/branch-A-Assertions

Add assertions

Ashuh committed:
Merge branch 'master' into branch-A-CodeQuality

Ashuh committed:
Merge pull request #2 from Ashuh/branch-A-CodeQuality

Delete old text-based UI code

Ashuh committed:
Set up CI

Ashuh committed:
Delete old tests

Ashuh committed:
Extract parsing logic into Parser class

Parsing logic is implemented as private methods in the Duke class.

Parsing logic may also be useful in places other than the Duke class.

Let's extract the parseDateTime() and parseDuration() methods from Duke into a new Parser class and make them static methods.

Extracting the parsing methods and making them static allows us to easily make use of them from any class.

Ashuh committed:
Add support for editing Tasks by right-clicking their respective cards

Ashuh committed:
Merge branch 'branch-C-Update'

Ashuh committed:
Add Ui.png

Ashuh committed:
Set theme jekyll-theme-midnight

Ashuh committed:
Update user guide

Ashuh committed:
Merge branch 'master' of https://github.com/Ashuh/ip

Ashuh committed:
Add tests for Parser

Ashuh committed:
Set handler methods in TaskCard to private

Ashuh committed:
Change exception name

Ashuh committed:
Extract Task parsing logic in Storage into new method

Ashuh committed:
Reduce nesting in Storage

Ashuh committed:
Set theme jekyll-theme-cayman

Ashuh committed:
Merge branch 'master' of https://github.com/Ashuh/ip

Ashuh committed:
Add line breaks in user guide

Ashuh committed:
Update mainClassName in build.gradle
